File: impl-go/no_0146_lru_cache/no_0146_lru_cache.go
package no_0146_lru_cache

import (
	"container/list"
)

type Item struct {
	key   int
	value int
}

type LRUCache struct {
	m        map[int]*list.Element
	l        *list.List
	capacity int
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		m:        make(map[int]*list.Element, capacity),
		l:        list.New(),
		capacity: capacity,
	}
}

func (this *LRUCache) Get(key int) int {
	ele := this.m[key]
	if ele == nil {
		return -1
	}
	this.l.MoveToFront(ele)
	return ele.Value.(*Item).value
}

func (this *LRUCache) Put(key int, value int) {
	ele := this.m[key]
	if ele != nil {
		ele.Value.(*Item).value = value
		this.l.MoveToFront(ele)
		return
	}
	for len(this.m) >= this.capacity {
		back := this.l.Back()
		this.l.Remove(back)
		delete(this.m, back.Value.(*Item).key)
	}
	ele = this.l.PushFront(&Item{key: key, value: value})
	this.m[key] = ele
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */

File: impl-rust/src/no_0146_lru_cache.rs
use std::collections::HashMap;
use std::collections::LinkedList;

struct LRUCache {
    map: HashMap<i32, i32>,
    list: LinkedList<i32>,
    capacity: i32,
}


/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 * 
 * Your LRUCache object will be instantiated and called as such:
 * let obj = LRUCache::new(capacity);
 * let ret_1: i32 = obj.get(key);
 * obj.put(key, value);
 */
impl LRUCache {

    fn new(capacity: i32) -> Self {
        LRUCache{
            map: HashMap::with_capacity(capacity as usize),
            list: LinkedList::new(),
            capacity,
        }
    }
    
    fn get(&self, key: i32) -> i32 {
        let res = self.map.get(&key);
        if res.is_some() {
            // 将 item 移到前面
            // todo: 这里怎么获得 item 的指针呢？
        }
        unimplemented!()
    }
    
    fn put(&self, key: i32, value: i32) {
        unimplemented!()
    }
}
